Add plugin: Wordflow Tracker

LeCheenaX opened the pull request:
# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/LeCheenaX/WordFlow-Tracker/

## Release Checklist
- [x] I have tested the plugin on
  - [x]  Windows
  - [x]  macOS
  - [x]  Linux
  - [x]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

LeCheenaX commented:
> Thank you for your submission, an automated scan of your plugin code's revealed the following issues:
> 
> ### Required
> [[1](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L13-L23)][[2](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L282-L297)][[3](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L299-L446)]:Remember to rename the sample classes to something that makes sense to your plugin. If there are any errors, Obsidian will spit out this class name, so this will be important for users to debug if they need to.
> 
> [[1](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L82)]:Please remove anything from the sample plugin that you don't need.
> 
> [[1](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L440)][[2](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L441)][[3](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L442)][[4](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L443)][[5](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L444)]:You should avoid assigning styles via JavaScript or in HTML and instead move all these styles into CSS so that they are more easily adaptable by themes and snippets.
> 
> [[1](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/recorder.ts#L19)][[2](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L77)][[3](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L235)][[4](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L33)][[5](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L74)][[6](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L86)][[7](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L89)][[8](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L105)][[9](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L106)][[10](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L124)][[11](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L125)][[12](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L140)][[13](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L156)][[14](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L185)][[15](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/DocTracker.ts#L195)]:You should consider limiting the number of `console.log`s in your code, to not pollute the dev console.
> 
> ### Optional
> [[1](https://github.com/LeCheenaX/WordFlow-Tracker/blob/b137d12fb051bb846a216696eba4ecb61b05cbdc/src/main.ts#L65)]:Casting to `any` should be avoided as much as possible.
> 
> Do NOT open a new PR for re-validation. Once you have pushed all of the required changes to your repo, the bot will update the labels on this PR within 6 hours. If you think some of the required changes are incorrect, please comment with `/skip` and the reason why you think the results are incorrect.

@ObsidianReviewBot Thank you for the review. The mentioned issues have been fixed now and pending on your review again. Thanks.

LeCheenaX reviewed (commented):
Should it be fine now for a human review? @ObsidianReviewBot

LeCheenaX reviewed (commented):
Update plugin description.

LeCheenaX commented:
@joethei Sorry for the interruption. It's observed that some pull requests after this post has been assigned a human for manual reviewing. May I check if it's appropriate to assign someone at the next availability?

joethei commented:
[**Warning: 1.2.0+ versions are very buggy. As multiple classes are rebuilt, please stay in 1.1.1 version for a while until the bugs are fixed.**](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/README.md#L4)
If you know that these versions are buggy then you should undo the releases and work on fixing the bugs.
Most mxisting users won't read a message in the README if they update the plugin via the app.


["fundingUrl": "to-be-continued",](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/manifest.json#L9)
This should be empty / omitted if you don't have a link that fits here.

[.setName('Current Recorder')](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L471), [contentEl.createEl("h3", { text: "Rename Recorder" });](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L974)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)


[containerEl.createEl('h3', {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L539-L542), [new Setting(container).setName('Recording contents setting').setHeading();](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L682)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

[new Setting(container).setName('Recording options').setHeading();](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L847)
[Avoid "options" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

[containerEl.createEl('h3', {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L539-L542)
For section headings in settings use:
```ts
new Setting(containerEl).setName('name here').setHeading();
```

[public async extractData(noteContent: string): Promise< Map<string, ExistingData>> {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/MetaDataParser.ts#L25-L156)
Use the `MetadataCache` API to read note frontmatter.

[import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/TableParser.ts#L2), [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/ListParser.ts#L2), [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/DataRecorder.ts#L4)
Import moment from the Obsidian package instead of shipping your own copy.

LeCheenaX commented:
> [**Warning: 1.2.0+ versions are very buggy. As multiple classes are rebuilt, please stay in 1.1.1 version for a while until the bugs are fixed.**](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/README.md#L4) If you know that these versions are buggy then you should undo the releases and work on fixing the bugs. Most mxisting users won't read a message in the README if they update the plugin via the app.
> 
> ["fundingUrl": "to-be-continued",](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/manifest.json#L9) This should be empty / omitted if you don't have a link that fits here.
> 
> [.setName('Current Recorder')](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L471), [contentEl.createEl("h3", { text: "Rename Recorder" });](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L974) [Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)
> 
> [containerEl.createEl('h3', {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L539-L542), [new Setting(container).setName('Recording contents setting').setHeading();](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L682) [Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)
> 
> [new Setting(container).setName('Recording options').setHeading();](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L847) [Avoid "options" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)
> 
> [containerEl.createEl('h3', {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L539-L542) For section headings in settings use:
> 
> ```ts
> new Setting(containerEl).setName('name here').setHeading();
> ```
> 
> [public async extractData(noteContent: string): Promise< Map<string, ExistingData>> {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/MetaDataParser.ts#L25-L156) Use the `MetadataCache` API to read note frontmatter.
> 
> [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/TableParser.ts#L2), [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/ListParser.ts#L2), [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/DataRecorder.ts#L4) Import moment from the Obsidian package instead of shipping your own copy.

**Sorry** for my late reply. I really appreciate your detailed perception regarding these improvements on my plugin. Indeed I missed many points for imprudence and I accepts all the advice above. Below is the current progress I had implemented:
- [x] Deleted the funding url as I don't need it at the point. 
> ["fundingUrl": "to-be-continued",](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/manifest.json#L9)
This should be empty / omitted if you don't have a link that fits here.

- [x] Change "Rename Recorder" to "Rename recorder" to implement the sentence case. 
> [.setName('Current Recorder')](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L471), [contentEl.createEl("h3", { text: "Rename Recorder" });](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L974)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

- [x] Deleted the word "setting" from the setting name, current name is: "Recording contents".
> [containerEl.createEl('h3', {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L539-L542), [new Setting(container).setName('Recording contents setting').setHeading();](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L682)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

- [x] Deleted the word "option" from the setting name, current name is: "Recording miscellaneous". 
> [new Setting(container).setName('Recording options').setHeading();](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L847) [Avoid "options" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

- [x] Replace the unfavorred 'h3' heading creation with the 'new Setting()' implementation
> [containerEl.createEl('h3', {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/main.ts#L539-L542) For section headings in settings use:
> ```ts
> new Setting(containerEl).setName('name here').setHeading();
> ```

- [x] Importing 'moment' from Obsidian package rather than from other package. 
> [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/TableParser.ts#L2), [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/ListParser.ts#L2), [import moment from 'moment';](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/DataRecorder.ts#L4) Import moment from the Obsidian package instead of shipping your own copy.

**You can check [this commit](https://github.com/LeCheenaX/WordFlow-Tracker/commit/8775cf0a84591e44015789742739717131cdfd2d) to view the above changes.** 

And the **WIP changes** are as follows:
- [ ] Fixed the functional bugs of 1.20+ versions. 
    - [x] 1.20+ releases are withdrawed and set as pre-release. Current latest release is 1.1.1 
> [Warning: 1.2.0+ versions are very buggy. As multiple classes are rebuilt, please stay in 1.1.1 version for a while until the bugs are fixed.](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/README.md#L4)
If you know that these versions are buggy then you should undo the releases and work on fixing the bugs.
Most mxisting users won't read a message in the README if they update the plugin via the app.

- [ ] Migrating from my implementaion to the Obsidian metadataCache method
> [public async extractData(noteContent: string): Promise< Map<string, ExistingData>> {](https://github.com/LeCheenaX/WordFlow-Tracker/blob/2a671a0a6343cc056d3af2c10b2de909c5c6839e/src/MetaDataParser.ts#L25-L156)
Use the MetadataCache API to read note frontmatter.

Will let you know by replying under this post later once I finished the adjustments. Warm regards to your kind reviewing : )
